src/apps/bm_devkit/serial_payload_example/user_code/user_code.cpp

```diff
@@ -58,12 +116,19 @@ void setup(void) {
   PLUART::setUseLineBuffer(true);
   // Set a line termination character per protocol of the sensor.
   PLUART::setTerminationCharacter((char)line_term_config);
+
+  // Setup PLUART transactions as needed per serial type
   if (uart_mode_config == MODE_RS485_HD) {
     printf("Enabling RS485 Half-Duplex.\n");
     bristlefin.setRS485HalfDuplex();  // Set the RS485 transceiver to Half-Duplex.
     // Set up PLUART transactions for proper RS485 enable Tx / enable Rx
     PLUART::enableTransactions(Bristlefin::setRS485Tx, Bristlefin::setRS485Rx);
+  } else if (uart_mode_config == MODE_SDI12) {
+    printf("Enabling SDI-12 serial.\n");
+    // Set up PLUART transactions for proper SDI12 enable Tx / enable Rx
+    PLUART::enableTransactions(sdi_break_mark, Bristlefin::sdi12Rx);
   }
```

Comment:
Since all 3 of the other hardware control functions here are in the `Bristlefin` namespace, the first thing that occurs to me is that maybe it makes sense to move `sdi_break_mark` there as well.

However, all those other methods just involve calling `IOWrite` on various pins, whereas `sdi_break_mark` is entirely a sequence of `PLUART` calls with some additional timing. The same is true of `sdi_wake`.

The `Bristlefin` nameespace doesn't know about `PLUART`, and these SDI-12 sequences require too many calls to shoehorn it in there with some callbacks or duck typing. Putting the SDI-12 functions in the `Bristlefin` namespace would required adding a lot of dependencies that are otherwise unused there. That's a pretty clear sign that we should design it differently.

I think it makes the most sense to move `sdi_wake` and `sdi_break_mark` into the `PLUART` namespace. I'm pretty sure all the necessary dependencies are already there. That's my recommendation.

Reply:
@towynlin `sdi_break_mark` needs bristlefin.h. Do we want to add bristlefin to PLUART? A bristlefin board always has a Mote with a PLUART, but a Mote with a PLUART isn't always connected to a bristlefin board.

Or, do we put these SDI12-specific sequence functions in PLUART, but then have the caller responsible for passing in the SDI12-tx-enable function per whatever board lib it's using. Imagining we make another adapter board some day with an SDI-12 transceiver on it, this would require less copy-pasting than putting stuff in bristlefin.h. However, we're then bloating PLUART with code that isn't relevant to most systems that have a PLUART. All systems importing this file will have a Mote with a PLUART, but many of them won't have an SDI-12 transceiver. If we want to avoid that, we could alos just introduce an sdi12.h file.

Reply:
@evanShap  Ah! I didn't realize that PLUART didn't already have access to Bristlefin. Well, derp. Probably then a separate `sdi12.{c,h}` would be best then actually.

Reply:
Yea, I'm coming around to separate sdi12 and rs485 files being the most future-proof approach. Given that, a couple more questions:
- What is the preference style guide for implementing these - inlines in a single .h, implement in .c, implement in .cpp?
- I like the idea of including and wrapping these in bristlefin.h, rather than separately including these and bristlefin.h in the app. Reasoning:
  - Bristlefin HW _always_ has access to SDI12 and RS485, (even though a particular app might not use them).
  - This minimizes/simplifies the app code, which I think is pretty important given the level of experience of most of the users.
  - This still provides a clean + clear path for implementing different mote interface boards with some subset of RS485 and/or SDI-12:
    - 1. write a header for your carrier board, like bristlefin.h, but with your custom stuff.
    - 2. include and write wrappers for sdi12/rs485 as needed, but passing in the hooks for your carrier boards enable GPIOs.
    - 3. In your app, setup the PLUART transaction identically to how you do using bristlefin, just pass in a handle to your carrier board instead of the bristlefin.
@towynlin thoughts?

Reply:
Love it. 👍 Let the .h file be pretty minimal — just declare your public functions and any types that consuming code might need, then in a .cpp file write your implementations.